In [11]:
import pandas as pd

In [12]:
# hotels = pd.read_csv('hotels.csv')

hotels = pd.read_csv('hotels_new.csv')

hotels.head(3)

average_score  total_number_of_reviews  \
0            8.4                -0.057096   
1            8.3                -0.315253   
2            8.9                -0.704731   

   total_number_of_reviews_reviewer_has_given  reviewer_score  month  \
0                                    0.571429            10.0      2   
1                                    1.571429             6.3      1   
2                                    1.571429             7.5     10   

   tag_Couple  tag_Submitted_from_a_mobile_device  tag_Business_trip  \
0        True                                True              False   
1        True                               False               True   
2       False                                True              False   

   tag_Standard_Double_Room  tag_Stayed_1_night  ...  \
0                     False               False  ...   
1                      True                True  ...   
2                     False               False  ...   

   hotel_country_Netherlands  hotel_country_Austria  hotel_country_Spain  \
0                          0                      0                    0   
1                          0                      0                    0   
2                          0                      0                    0   

   subjectivity  reviewer_nationality_United Kingdom  \
0          True                                    1   
1          True                                    1   
2         False                                    0   

   reviewer_nationality_Other  reviewer_nationality_United States of America  \
0                           0                                              0   
1                           0                                              0   
2                           1                                              0   

   reviewer_nationality_Australia  No_Negative  No_Positive  
0                               0        False        False  
1                               0        False        False  
2                               0        False         True  

[3 rows x 28 columns]

In [13]:
# hotels.drop(axis=1, columns=['hotel_address', 'review_date', 'hotel_name', \
#    'reviewer_nationality', 'negative_review', 'positive_review', \
#        'tags', 'days_since_review'], inplace=True)

In [14]:
# hotels.fillna(0,inplace=True)

In [15]:
# Разбиваем датафрейм на части, необходимые для обучения и тестирования модели  
# Х - данные с информацией об отелях, у - целевая переменная (рейтинги отелей)  
X = hotels.drop(['reviewer_score'], axis = 1)  
y = hotels['reviewer_score'] 

In [16]:
# Загружаем специальный инструмент для разбивки:  
from sklearn.model_selection import train_test_split  

In [17]:
# Наборы данных с меткой "train" будут использоваться для обучения модели, "test" - для тестирования.  
# Для тестирования мы будем использовать 25% от исходного датасета.  
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [18]:
# Импортируем необходимые библиотеки:  
from sklearn.ensemble import RandomForestRegressor # инструмент для создания и обучения модели  
from sklearn import metrics # инструменты для оценки точности модели  
  
# Создаём модель  
regr = RandomForestRegressor(n_estimators=100)  
      
# Обучаем модель на тестовом наборе данных  
regr.fit(X_train, y_train)  
      
# Используем обученную модель для предсказания рейтинга отелей в тестовой выборке.  
# Предсказанные значения записываем в переменную y_pred  
y_pred = regr.predict(X_test)  


In [19]:
# Сравниваем предсказанные значения (y_pred) с реальными (y_test), и смотрим насколько они отличаются  
# Метрика называется Mean Absolute Percentage Error (MAPE) и показывает среднюю абсолютную процентную ошибку предсказанных значений от фактических.  
print('MAPE:', metrics.mean_absolute_percentage_error(y_test, y_pred))

MAPE: 0.16100001937816935


1 MAPE: 0.1413364989016283

2 MAPE: 0.13434831921902826

3 MAPE: 0.13324687641304117

4 MAPE: 0.13364068838984453

5 MAPE: 0.13314658276790986

6 MAPE: 0.1321199161523903

7 MAPE: 0.16100001937816935

так отразились изменения в review, на kaggle они не попадут

Небольшой бонус:


In [20]:
# # убираем признаки которые еще не успели обработать, 
# # модель на признаках с dtypes "object" обучаться не будет, просто выберим их и удалим
# object_columns = [s for s in hotels.columns if hotels[s].dtypes == 'object']
# hotels.drop(object_columns, axis = 1, inplace=True)

# # заполняем пропуски самым простым способом
# hotels = hotels.fillna(0)